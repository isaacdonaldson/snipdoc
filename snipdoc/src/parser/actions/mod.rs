#[cfg(feature = "exec")]
pub mod exec;
