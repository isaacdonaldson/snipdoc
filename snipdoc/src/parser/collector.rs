use std::{
    collections::BTreeMap,
    path::{Path, PathBuf},
    str::FromStr,
};

use pest::{iterators::Pairs, Parser};
use rayon::prelude::*;
use serde::{Deserialize, Serialize};

use super::{html_tag, Rule};
use crate::{
    errors::ParserResult,
    parser::{SnippetKind, SnippetParse},
    read_file::RFile,
    walk::Walk,
};

/// A struct representing a snippet extracted from code
#[derive(Debug, Serialize, Deserialize)]
pub struct CollectSnippet {
    /// ID of the snippet.
    pub id: String,
    /// Defined if `inject` attribute exists in the snippet.
    pub inject_from: Option<SnippetKind>,
    /// Collect if `strip_prefix` attribute if exists exists in the snippet.
    // pub strip_prefix: Option<String>,
    /// Defined the tag open value of the snippet.
    pub tag_open: String,
    /// Defined the the tag close value of the snippet.
    pub tag_close: String,
    /// Hold all the line content inside the snippet.
    pub snippet: Vec<String>,
}

pub struct Collector<'a> {
    pub input: &'a str,
}

/// A struct for collecting snippets from files within a folder.
#[derive(Debug, Serialize, Deserialize)]
pub struct CollectSnippetsResults {
    pub root_folder: PathBuf,
    pub snippets: BTreeMap<PathBuf, Vec<CollectSnippet>>,
}

impl<'a> Collector<'a> {
    /// Constructs a new [`ParseFile`] with the provided input.
    #[must_use]
    pub const fn new(input: &'a str) -> Self {
        Self { input }
    }

    /// Constructs a `Collector` instance by collecting snippets from files
    /// within the provided `Walk`.
    #[must_use]
    pub fn walk(walk: &Walk) -> CollectSnippetsResults {
        let files = walk.get_files();

        tracing::debug!(
            count_files = files.len(),
            "start collect snippets from code"
        );
        let snippets = files
            .par_iter()
            .flat_map(|path| Self::file(path.as_path()).map(|findings| (path.clone(), findings)))
            .collect::<BTreeMap<_, _>>();

        CollectSnippetsResults {
            root_folder: walk.folder.clone(),
            snippets,
        }
    }

    // Processes a single file and extracts [`Vec<CollectSnippet>`].
    ///
    /// # Returns
    ///
    /// Returns `Some` containing the collected snippets if successful,
    /// otherwise returns `None`.
    #[allow(clippy::missing_errors_doc)]
    pub fn file(path: &Path) -> ParserResult<'_, Vec<CollectSnippet>> {
        let span = tracing::info_span!("parse_file", path = %path.display());
        let _guard = span.enter();

        let r_file = RFile::new(path)?;

        Collector::new(&r_file.content).run()
    }

    /// Parses the input file content and extracts snippets.
    ///
    /// # Errors
    ///
    /// This function may return an error if it fails to parse the input file.
    /// Other errors encountered during parsing will be logged.
    fn run(&self) -> ParserResult<'_, Vec<CollectSnippet>> {
        let pairs = SnippetParse::parse(Rule::file, self.input)?;

        let mut findings: Vec<CollectSnippet> = vec![];
        Self::collect_snippets(pairs, &mut findings);
        Ok(findings)
    }

    /// Recursively collects snippets from the provided pairs and populates the
    /// given vector with the snippets.
    ///
    /// This function recursively traverses the pairs, extracting snippets and
    /// their attributes.
    ///
    /// # Panics
    ///
    /// This function assumes that the parsing configuration always captures a
    /// snippet containing a tag open. If this assumption is violated, it
    /// indicates a misconfiguration or a critical issue in the parser's
    /// behavior. Consequently, in production code, encountering this panic
    /// indicates a severe problem that requires immediate attention.
    /// In testing scenarios, this panic should be captured to ensure the
    /// correctness of the parser.
    fn collect_snippets(pairs: Pairs<'_, Rule>, snippets: &mut Vec<CollectSnippet>) {
        if pairs.len() == 0 {
            return;
        }

        for pair in pairs {
            let inner = pair.clone().into_inner();

            match pair.as_rule() {
                Rule::snippet => {
                    let children: Pairs<'_, Rule> = pair.clone().into_inner();
                    let tag_open = html_tag::get_tag_open(&children);
                    let tag_close = html_tag::get_tag_close(children.clone());

                    let attributes = match html_tag::get_tag_attributes(tag_open) {
                        Ok(attributes) => attributes,
                        Err(err) => {
                            tracing::debug!(tag_open, err = %err, "could not extract attributes from the tag");
                            continue;
                        }
                    };

                    tracing::debug!(
                        tag_open,
                        tag_close,
                        attributes = format!("{:#?}", attributes),
                        "found attributes"
                    );

                    let mut lines = pair
                        .as_str()
                        .lines()
                        .map(std::string::ToString::to_string)
                        .skip(1)
                        .collect::<Vec<_>>();

                    lines.pop();

                    if let Some(last) = lines.last() {
                        if last == &tag_close.replace('\n', "") {
                            // TODO:: replace \n to crate::LINE_ENDING
                            lines.pop();
                        }
                    }

                    snippets.push(CollectSnippet {
                        // Attribute ID as part of the parser configuration is
                        // mandatory. the snippet should't be captured if id
                        // element is not present. In this case
                        // user `expect` should brake the parser.
                        id: attributes
                            .get("id")
                            .expect("assertion fails, snippet without element id")
                            .to_string(),
                        inject_from: attributes
                            .get("inject_from")
                            .and_then(|k| SnippetKind::from_str(k).ok()),
                        tag_open: tag_open.to_string(),
                        tag_close: tag_close.to_string(),
                        snippet: lines,
                    });

                    Self::collect_snippets(children, snippets);
                }
                _ => {
                    Self::collect_snippets(inner.clone(), snippets);
                }
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use insta::assert_debug_snapshot;

    use super::*;

    #[test]
    fn can_collect() {
        let content = r#"# Snipdoc
# parse the snippet with `inject_from` attribute
<!-- <snip id="description" inject_from="code"> -->
<!-- </snip> -->

# parse snippet with html comment
<!-- <snip id="installation"> -->
$ cargo install snipdoc
$ ssnipdoc --version
<!-- </snip> -->

# parse snippet without any spaces between the comment
<!--<snip id="no-spaces">-->
$ cargo install snipdoc
$ ssnipdoc --version
<!--</snip>-->

# parse snippet with double slash comment
// <snip id="double-slash">
double-slash
// </snip>

# parse snippet with triple slash comment
/// <snip id="triple-slash">
triple-slash
/// </snip>
///
# parse snippet with triple hashtag comment
# <snip id="hashtag">
hashtag
# </snip>

# Inner snippets
<!-- <snip id="level-1" -->
Level 1
// <snip id="level-2">
Level 2
// <snip id="level-3">
Level 3
// </snip>
// </snip>
<!-- </snip> -->
"#;

        let collector = Collector::new(content);
        assert_debug_snapshot!(collector.run());
    }
}
