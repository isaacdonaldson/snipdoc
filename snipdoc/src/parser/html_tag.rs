use std::collections::BTreeMap;

use pest::iterators::Pairs;
use scraper::{Html, Selector};

use super::Rule;
use crate::errors::{ParseError, ParserResult};

/// Extracts the tag open element from the captured pair.
///
/// # Panics
///
/// This function assumes that the parsing configuration always captures a
/// snippet containing a tag open. If this assumption is violated, it
/// indicates a misconfiguration or a critical issue in the parser's
/// behavior. Consequently, in production code, encountering this panic
/// indicates a severe problem that requires immediate attention.
/// In testing scenarios, this panic should be captured to ensure the
/// correctness of the parser.
#[must_use]
pub fn get_tag_open<'b>(pair_children: &'b Pairs<'_, Rule>) -> &'b str {
    let pair = pair_children
        .clone()
        .next()
        .expect("assertion fails, snippet without tag open");

    assert!(
        pair.as_rule() == Rule::tag_open,
        "Expected tag_open rule, found {:?}",
        pair.as_rule()
    );

    pair.into_inner()
        .nth(1)
        .expect("Expected at least two")
        .as_str()
}

/// Extracts the tag close element from the captured pair.
///
///
/// This function iterates over the provided `pair_children` iterator to
/// find the tag close element. It assumes that the captured pairs
/// represent a snippet, and it searches for the tag close element
/// within these pairs.
///
/// # Panics
///
/// This function panics if the tag close element is not found in the
/// captured pairs. In production code, encountering this panic
/// indicates a severe problem, such as a misconfiguration or a critical
/// issue in the parser's behavior, which requires immediate attention.
/// In testing scenarios, this panic should be captured to ensure the
/// correctness of the parser.
#[must_use]
pub fn get_tag_close(pair_children: Pairs<'_, Rule>) -> &'_ str {
    for x in pair_children {
        match x.as_rule() {
            Rule::tag_close => {
                return x.as_str();
            }
            _ => continue,
        }
    }
    panic!("tag close not found")
}

/// Extracts the attributes from the given HTML tag.
///
/// For Example:
/// * The given tag is `<snip id="quick_start">` the result will be {"id":
///   "`quick_start`"}
///
/// # Errors
///
/// This function returns an error in the following cases:
///
/// * The tag is not of type `snipdoc`.
/// * The tag cannot be parsed.
pub fn get_tag_attributes(tag: &str) -> ParserResult<'_, BTreeMap<String, String>> {
    let html = Html::parse_fragment(tag);

    let selector_name = "snip";
    let selector = Selector::parse(selector_name)?;
    let attributes = html
        .select(&selector)
        .next()
        .ok_or_else(|| ParseError::SelectorNotFound {
            selector: selector_name.to_string(),
            tag: tag.to_string(),
        })?
        .value()
        .attrs();

    Ok(attributes
        .map(|(key, value)| (key.to_string(), value.to_string()))
        .collect())
}

/// Extracts the comment tag from the captured pair.
///
/// for the given tag: <!-- <snip id="SNIPPET_ID"> the results will be result of
/// this function is `<!--`
///
/// # Panics
///
/// This function assumes that the parsing configuration always captures a
/// snippet containing a tag open. If this assumption is violated, it
/// indicates a misconfiguration or a critical issue in the parser's
/// behavior. Consequently, in production code, encountering this panic
/// indicates a severe problem that requires immediate attention.
/// In testing scenarios, this panic should be captured to ensure the
/// correctness of the parser.
#[must_use]
pub fn get_comment_tag_open<'b>(pair_children: &'b Pairs<'_, Rule>) -> &'b str {
    let pair = pair_children
        .clone()
        .next()
        .expect("assertion fails, snippet without tag open");

    assert!(
        pair.as_rule() == Rule::tag_open,
        "Expected tag_open rule, found {:?}",
        pair.as_rule()
    );

    pair.into_inner()
        .next()
        .expect("Expected at least two")
        .as_str()
}

/// Extracts the close comment tag from the captured pair.
///
/// for the given tag: <!-- <snip id="SNIPPET_ID"> --> the results will be
/// result of this function is `-->`
///
///
/// # Panics
///
/// This function assumes that the parsing configuration always captures a
/// snippet containing the  comment close of the tag. If this assumption is
/// violated, it indicates a misconfiguration or a critical issue in the
/// parser's behavior. Consequently, in production code, encountering this panic
/// indicates a severe problem that requires immediate attention.
/// In testing scenarios, this panic should be captured to ensure the
/// correctness of the parser.
#[must_use]
pub fn get_comment_tag_of_tag_open<'b>(pair_children: &'b Pairs<'_, Rule>) -> Option<&'b str> {
    let pair = pair_children
        .clone()
        .next()
        .expect("assertion fails, snippet without tag open");

    assert!(
        pair.as_rule() == Rule::tag_open,
        "Expected tag_open rule, found {:?}",
        pair.as_rule()
    );

    pair.into_inner().nth(2).map(|p| p.as_str())
}
