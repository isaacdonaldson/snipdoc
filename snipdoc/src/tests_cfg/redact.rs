#[cfg(windows)]
pub const REGEX_REPLACE_LINE_ENDING: &str = r"\\r\\n";

#[cfg(not(windows))]
pub const REGEX_REPLACE_LINE_ENDING: &str = r"\\n";

#[must_use]
pub fn all() -> Vec<(&'static str, &'static str)> {
    vec![(REGEX_REPLACE_LINE_ENDING, "[NEW_LINE]")]
}
