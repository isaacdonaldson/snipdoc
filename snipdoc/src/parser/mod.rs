mod actions;
pub mod collector;
mod html_tag;
pub mod injector;

use std::{path::PathBuf, str::FromStr};

use pest_derive::Parser;
use serde::{Deserialize, Serialize};

#[cfg(feature = "exec")]
use crate::parser::actions::exec;

#[derive(Parser)]
#[grammar = "snippet.pest"]
pub struct SnippetParse;

#[derive(Serialize, Deserialize, Debug)]
pub struct Snippet {
    pub id: String,
    pub content: String,
    pub kind: SnippetKind,
    pub path: PathBuf,
}

#[cfg_attr(feature = "cli", derive(clap::ValueEnum))]
#[derive(Default, Clone, Serialize, Deserialize, Debug, PartialEq, Eq)]
pub enum SnippetKind {
    Yaml,
    Code,
    #[default]
    Any,
}

impl FromStr for SnippetKind {
    type Err = ();

    fn from_str(input: &str) -> std::result::Result<Self, ()> {
        match input {
            "yaml" => Ok(Self::Yaml),
            "code" => Ok(Self::Code),
            "any" => Ok(Self::Any),
            _ => Err(()),
        }
    }
}

impl Snippet {
    /// Returns the snippet content, filtered based on `strip_prefix` if
    /// specified.
    #[must_use]
    pub fn create_content(&self, inject_actions: &injector::InjectContentAction) -> String {
        #[cfg(feature = "exec")]
        let content = if inject_actions.kind == injector::InjectAction::Exec {
            exec::run(&self.content).unwrap_or_else(|err| {
                tracing::error!(err, "execute snippet command failed");
                self.content.to_string()
            })
        } else {
            self.content.to_string()
        };

        #[cfg(not(feature = "exec"))]
        let content = self.content.to_string();

        let content = inject_actions
            .template
            .before_inject(&content, &inject_actions.kind);

        let content = content
            .lines()
            .filter_map(|line| {
                // validate if i can remove this code
                if line.contains("<snip") || line.contains("</snip") {
                    return None;
                }
                let line = inject_actions.strip_prefix.as_ref().map_or_else(
                    || line.to_string(),
                    |prefix_inject| line.strip_prefix(prefix_inject).unwrap_or(line).to_string(),
                );

                if let Some(add_prefix) = &inject_actions.add_prefix {
                    Some(format!("{add_prefix}{line}"))
                } else {
                    Some(line)
                }
            })
            .collect::<Vec<_>>()
            .join(crate::LINE_ENDING);

        inject_actions
            .template
            .after_inject(&content, &inject_actions.kind)
    }
}

#[cfg(test)]
mod tests {
    use insta::{assert_debug_snapshot, with_settings};

    use super::*;
    use crate::{
        parser::injector::{InjectAction, Template},
        tests_cfg,
    };

    #[test]
    fn can_get_snippet_content_without_action() {
        let snippet = tests_cfg::get_snippet();

        let action = injector::InjectContentAction {
            kind: InjectAction::Copy,
            snippet_id: "id".to_string(),
            inject_from: SnippetKind::Any,
            strip_prefix: None,
            add_prefix: None,
            template: Template::default(),
        };

        with_settings!({filters => tests_cfg::redact::all()}, {
            assert_debug_snapshot!(snippet.create_content(&action));
        });
    }

    #[test]
    fn can_get_snippet_content_with_template_action() {
        let snippet = tests_cfg::get_snippet();

        let action = injector::InjectContentAction {
            kind: InjectAction::Copy,
            snippet_id: "id".to_string(),
            inject_from: SnippetKind::Any,
            strip_prefix: None,
            add_prefix: None,
            template: Template::new("```sh\n{snippet}\n```"),
        };

        with_settings!({filters => tests_cfg::redact::all()}, {
            assert_debug_snapshot!(snippet.create_content(&action));
        });
    }

    #[test]
    fn can_get_snippet_content_with_strip_prefix_action() {
        let snippet = tests_cfg::get_snippet();

        let action = injector::InjectContentAction {
            kind: InjectAction::Copy,
            snippet_id: "id".to_string(),
            inject_from: SnippetKind::Any,
            strip_prefix: Some("$ ".to_string()),
            add_prefix: None,
            template: Template::default(),
        };

        with_settings!({filters => tests_cfg::redact::all()}, {
            assert_debug_snapshot!(snippet.create_content(&action));
        });
    }

    #[test]
    fn can_get_snippet_content_with_add_prefix_action() {
        let snippet = tests_cfg::get_snippet();

        let action = injector::InjectContentAction {
            kind: InjectAction::Copy,
            snippet_id: "id".to_string(),
            inject_from: SnippetKind::Any,
            strip_prefix: None,
            add_prefix: Some("$".to_string()),
            template: Template::default(),
        };

        with_settings!({filters => tests_cfg::redact::all()}, {
            assert_debug_snapshot!(snippet.create_content(&action));
        });
    }

    #[test]
    fn can_get_snippet_content_with_combination_action() {
        let snippet = tests_cfg::get_snippet();

        let action = injector::InjectContentAction {
            kind: InjectAction::Copy,
            snippet_id: "id".to_string(),
            inject_from: SnippetKind::Any,
            strip_prefix: Some("$ ".to_string()),
            add_prefix: Some("- ".to_string()),
            template: Template::new("```sh\n{snippet}\n```"),
        };

        with_settings!({filters => tests_cfg::redact::all()}, {
            assert_debug_snapshot!(snippet.create_content(&action));
        });
    }

    #[cfg(feature = "exec")]
    #[test]
    fn can_get_snippet_with_exec_action_with_template() {
        let mut snippet = tests_cfg::get_snippet();
        snippet.content = r"echo calc result: $((1+1))".to_string();

        let action = injector::InjectContentAction {
            kind: InjectAction::Exec,
            snippet_id: "id".to_string(),
            inject_from: SnippetKind::Any,
            strip_prefix: None,
            add_prefix: None,
            template: Template::new("```sh\n{snippet}\n```"),
        };

        with_settings!({filters => tests_cfg::redact::all()}, {
            assert_debug_snapshot!(snippet.create_content(&action));
        });
    }
}
