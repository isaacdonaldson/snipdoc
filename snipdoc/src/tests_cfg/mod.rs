use std::{collections::BTreeMap, path::PathBuf};
pub mod redact;
use crate::parser::{collector::CollectSnippet, Snippet, SnippetKind, SnippetTemplate};

#[must_use]
pub fn get_collect_snippets() -> Vec<CollectSnippet> {
    vec![
        CollectSnippet {
            id: "description".to_string(),
            snippet: vec!["test".to_string(), "snipdoc".to_string()],
            inject_from: None,
            tag_open: "<snip id=\"description\">".to_string(),
            tag_close: "<!-- </snip> -->\n".to_string(),
        },
        CollectSnippet {
            id: "installation".to_string(),
            snippet: vec![
                "```".to_string(),
                "cargo install snipdoc".to_string(),
                "```".to_string(),
            ],
            inject_from: None,
            tag_open: "<snip id=\"install\">".to_string(),
            tag_close: "<!-- </snip> -->\n".to_string(),
        },
        CollectSnippet {
            id: "from-yaml".to_string(),
            snippet: vec!["ignore-snippet".to_string()],
            inject_from: Some(SnippetKind::Yaml),
            tag_open: "<snip id=\"from-yaml\">".to_string(),
            tag_close: "<!-- </snip> -->\n".to_string(),
        },
    ]
}

#[must_use]
pub fn get_custom_templates() -> BTreeMap<String, SnippetTemplate> {
    BTreeMap::from([
        (
            "CUSTOM_ID_1".to_string(),
            SnippetTemplate {
                content: "---> {snippet}".to_string(),
            },
        ),
        (
            "CUSTOM_ID_2".to_string(),
            SnippetTemplate {
                content: "({snippet})".to_string(),
            },
        ),
    ])
}

#[must_use]
pub fn get_snippet() -> Snippet {
    Snippet {
        id: "test".to_string(),
        content: "$ cargo install snipdoc\n$ snipdoc --version".to_string(),
        kind: SnippetKind::Code,
        path: PathBuf::from("main.rs"),
    }
}

#[must_use]
pub fn get_snippet_to_inject() -> BTreeMap<String, Snippet> {
    BTreeMap::from([
        (
            "description".to_string(),
            Snippet {
                id: "test".to_string(),
                content: "snipdoc".to_string(),
                kind: SnippetKind::Code,
                path: PathBuf::from("main.rs"),
            },
        ),
        (
            "installation".to_string(),
            Snippet {
                id: "test".to_string(),
                content: "$ cargo install snipdoc\n$ snipdoc --version".to_string(),
                kind: SnippetKind::Code,
                path: PathBuf::from("main.rs"),
            },
        ),
        (
            "inject_from_yaml".to_string(),
            Snippet {
                id: "test".to_string(),
                content: "inject_from_yaml".to_string(),
                kind: SnippetKind::Yaml,
                path: PathBuf::from("main.rs"),
            },
        ),
    ])
}
