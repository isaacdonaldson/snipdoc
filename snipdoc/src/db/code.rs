//! This module defines the `Code` database, which get list collected snippets
//! from live code and converts them to a common structure for use with
//! `snipdoc`.
use std::{collections::BTreeMap, path::PathBuf};

use super::{DBData, Db, Error, Result, Snippet};
use crate::parser::{collector::CollectSnippet, SnippetKind};

pub struct Code {
    pub snippets: BTreeMap<PathBuf, Vec<CollectSnippet>>,
}

impl Code {
    #[must_use]
    pub fn new(snippets: BTreeMap<PathBuf, Vec<CollectSnippet>>) -> Self {
        Self { snippets }
    }
}

impl Db for Code {
    fn load(&self) -> Result<'_, DBData> {
        let mut data = DBData::default();

        for (path, snippets) in &self.snippets {
            for snippet in snippets {
                // when loading snippets from the code, we should get only snippets and not
                // placeholder that we should inject
                if snippet.inject_from.is_none() {
                    data.snippets.insert(
                        snippet.id.clone(),
                        Snippet {
                            id: snippet.id.clone(),
                            content: snippet.snippet.join(crate::LINE_ENDING),
                            kind: SnippetKind::Code,
                            path: path.clone(),
                        },
                    );
                }
            }
        }

        Ok(data)
    }
    fn save(&self, _: &[&CollectSnippet]) -> Result<'_, ()> {
        Err(Error::NotSupported)
    }
}

#[cfg(test)]
mod tests {
    use insta::{assert_debug_snapshot, with_settings};

    use super::*;
    use crate::tests_cfg;

    #[test]
    fn can_load() {
        let code = Code {
            snippets: BTreeMap::from([(
                PathBuf::from("README.md"),
                tests_cfg::get_collect_snippets(),
            )]),
        };

        with_settings!({filters => vec![
            (tests_cfg::REGEX_REPLACE_LINE_ENDING, "\n")
        ]}, {
            assert_debug_snapshot!(code.load());
        });
    }

    #[test]
    fn can_save() {
        let code = Code {
            snippets: BTreeMap::new(),
        };

        let save_snippets = CollectSnippet {
            id: "description".to_string(),
            snippet: vec!["test".to_string(), "snipdoc".to_string()],
            inject_from: None,
            tag_open: "<snip id=\"description\">".to_string(),
            tag_close: "<!-- </snip> -->\n".to_string(),
        };

        with_settings!({filters => vec![
            (tests_cfg::REGEX_REPLACE_LINE_ENDING, "\n")
        ]}, {
            assert!(code.save(&[&save_snippets]).is_err());
        });
    }
}
