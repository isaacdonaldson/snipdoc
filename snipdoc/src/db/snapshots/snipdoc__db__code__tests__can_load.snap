---
source: snipdoc/src/db/code.rs
expression: code.load()
---
Ok(
    DBData {
        snippets: {
            "description": Snippet {
                id: "description",
                content: "test\nsnipdoc",
                kind: Code,
                path: "README.md",
            },
            "installation": Snippet {
                id: "installation",
                content: "```\ncargo install snipdoc\n```",
                kind: Code,
                path: "README.md",
            },
        },
    },
)
